### PR TITLE
Java implementation of tic tac toe (minimax, AB pruning, Spring MVC & React)

### PR DESCRIPTION
**Pull request by amucunguzi**
The code below can be roughly classified into three categories:

1. **Core Modules** (_com.nexttran.tictactoecore_ package): Somewhat of "business rules" of tic tac toe independent of UI. 
2. **Console Modules** (_com.nexttran.tictactoecore.console_  package): Tic tac toe classes for a console-based game. 
3. **Web Modules** (_com.nexttran.tictactoecore.web_ package): Tic tac toe classes for a web-based game. Using the Spring Framework for the backend. This package has two sub-packages:

- Spring MVC (_com.nexttran.tictactoecore.web.springmvc_ package): Spring is used for both the backend and the front-end. 

- Spring with React (_com.nexttran.tictactoecore.web.springwithreact_ package): React is used for the front-end.

That's the structure of the project (with a hierarchy). 

Unit tests follow the same packaging structure.

### CODE REVIEW

**Comment by ssunday** [comment]
Why do you have two versions of yourself in VCS? Er, like the github profile for the commits is the empty face and there's two of them? Maybe there's a mismatch between config users...

File: README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+### Unit & Integration Tests
+
+Unit & integration tests use JUnit (dependency in *pom.xml*). You can create a run/debbug configuration and execute all the tests in this projects at the same time. They should all pass.
+
+### How to Play the Game
+
+The easiest way to run a game is to use create an executable jar file. 
+
+As the project follows maven packaging structure, you can create jar files for all the three versions by running the following command in this project's root directory (make sure to have `mvn` on your machine):
```
Comment (ssunday) [inline README.md]: Perhaps a link to how to install mvn or what it is? (I know what it is, but it'd help helpful to say what version, etc.

File: reactapp/src/BoardView.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { Button } from 'reactstrap';
+import React from "react";
+
+const BoardView = (props) => (
+<div className="board" id="board">
```
Comment (ssunday) [inline reactapp/src/BoardView.js]: The indentation is a little off here. Since it's basically HTML, it should have the same indentation rules as HTML

File: reactapp/src/BoardView.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    <div className="board-row">
+    <Button size="lg" onClick={() => props.humanPlay(props.zero)}>{ props.zero }</Button>
```
Comment (ssunday) [inline reactapp/src/BoardView.js]: With these props/onclick buttons, could your iterate (map) through a list of them and make the buttons?

File: reactapp/src/App.test.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import App from './App';
+
+it('renders without crashing', () => {
```
Comment (cmvandrevala) [inline reactapp/src/App.test.js]: This test does not make any specific assertions, so it does not provide much in the way of useful feedback. Is at stands right now, this is almost like a smoke test rather than a unit test (i.e. the app works or it does not, but there is little explanation why). Is there a way that you can assert against something here in order to get better feedback?

File: reactapp/src/BoardComponent.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+import ComputerStart from "./ComputerStart";
+
+const AVAILABLE_SPOT = 1;
+
+class BoardComponent extends Component {
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Is this a scalable way to store the state of the board? Right now, you have literally hard-coded values like "zero", "one", "two",...

No need to address this comment right now, but this is something to think about for the future

File: reactapp/src/BoardComponent.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+computeFetch = async () => {
+
+    await fetch("http://localhost:8080/api/computer", {
+        method: 'PUT',
+        headers: {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+        },
+
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Not sure that you need these extra spaces. Consider implementing a linter like eslint

File: reactapp/src/BoardComponent.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        three: data.three, four: data.four, five: data.five,
+        six: data.six, seven: data.seven, eight: data.eight
+
+    }));
+}
+
+
+humanPlay = async (playedSpot) => {
+
+    await  this.setState({ error: "" }); // just to clear any previous errors
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Is there a way that you can express "just to clear any previous errors" without a comment? We generally like to steer of comments as much as possible.

File: reactapp/src/BoardComponent.js [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+    if (this.state.ongoing) {
+        await this.fetchHuman(playedSpot);
+        await this.updateGameState();
+
+        if (this.state.ongoing && !(this.state.error) &&
+            this.state.statusCode === AVAILABLE_SPOT) {
+
+
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Lots of extra spaces - watch your formatting.

File: reactapp/src/BoardComponent.js [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    }
+}
+
+
+fetchHuman = async (playedSpot) => {
+
+    if (this.state.ongoing) {
+
+        await fetch(`http://localhost:8080/api/human/${playedSpot}`, {
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Right now, you are hardcoding the URLs into your HTTP requests. This is fine until you have multiple environments (e.g., staging.tictactoe.com vs. tictactoe.com vs. localhost:8080). You might want to consider creating a config file that stores the base URL information

File: reactapp/src/BoardComponent.js [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+                two: this.state.two, three: this.state.three,
+                four: this.state.four, five: this.state.five,
+                six: this.state.six, seven: this.state.seven,
+                eight: this.state.eight})
+
+        }).then(response => {
+            if (response.ok) {
+            return response.json();
+        } else {
+            throw new Error("Hmm. Some unexpected behavior. Try to restart the game.");
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: I like that you are thinking about feedback for the user. In the future, it would be cool to see this expanded with specific error messages - assuming this is relevant to the stories that your mentors assign you :)

File: reactapp/src/BoardComponent.js [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+    } else {
+        this.setState({ending: "Hmm. The human won!"})
+    }
+}
+
+
+restartGame = () => {
+
+    this.setState({ zero: 0, one: 1, two: 2,
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Did you use integers to define the state when you first start the game? Or did you use strings? The state after you restart the game looks different than the state when you start a fresh game - maybe I am mistaken about that?

File: reactapp/src/BoardComponent.js [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    });
+}
+
+computerStart = async () => {
+    await this.restartGame();
+    await this.computerPlay();
+}
+
+getStatusCodeMeaning = (statusCode) => {
+    if (statusCode === 5) {
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Why are you using status codes here? Is it because JS does not support enums?

File: reactapp/src/BoardComponent.js [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+    <RestartGame restart={ this.restartGame }/>
+
+    <ComputerStart computerStart={ this.computerStart }/>
+
+    </Container>
+    </div>
+)
+}
+}
```
Comment (cmvandrevala) [inline reactapp/src/BoardComponent.js]: Try to format this like HTML

File: reactapp/src/BoardView.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+<div></div>
+<div className="board-row">
+    <Button size="lg" onClick={() => props.humanPlay(props.six)}>{ props.six }</Button>
+<Button size="lg" onClick={() => props.humanPlay(props.seven)}>{ props.seven }</Button>
+<Button size="lg" onClick={() => props.humanPlay(props.eight)}>{ props.eight} </Button>
+</div>
+</div>
+);
+
+export default BoardView;
```
Comment (cmvandrevala) [inline reactapp/src/BoardView.js]: Try to make sure that there is a single newline at the end of each file and there is no trailing whitespace after each line. You editor should have options to automatically set the newline and trim whitespace.

File: reactapp/src/GameEnding.js [hunk 15]
```diff
@@ -0,0 +1,5 @@
+import React from "react";
+
+const GameEnding  = (props) =>  (<div className="ending-message"><h4>{ props.ending }</h4></div>);
```
Comment (cmvandrevala) [inline reactapp/src/GameEnding.js]: Extra spaces

File: src/main/java/com/nexttran/tictactoecore/BoardTools.java [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+                && boardHasNonPlayedSpot(boardElements)) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean gameIsTie(List<String> boardElements) {
+        if (gameHasWinner(boardElements)
+                || boardHasNonPlayedSpot(boardElements)) return false;
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/BoardTools.java]: You don't need to return true or false directly. Instead:

`return !(gameHasWinner(boardElements) || boardHasNonPlayedSpot(boardElements))`

File: src/main/java/com/nexttran/tictactoecore/BoardTools.java [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+import java.util.List;
+
+
+public class BoardTools implements IBoardTools {
+
+    public String getElement(List<String> boardElements, int index) {
+        return boardElements.get(index);
+    }
+
+    public int getBoardSize(List<String> boardElements) {
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/BoardTools.java]: I am not sure why you need a board size here - doesn't your app hard code a 3x3 board above?

File: src/main/java/com/nexttran/tictactoecore/BoardTools.java [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+            }
+
+        }
+
+        return (gameHasWinner(boardElementsCopy)) ? secondSymbol : firstSymbol;
+    }
+
+    private static boolean isNumeric(String str)
+    {
+        return str.matches("-?\\d+(\\.\\d+)?");
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/BoardTools.java]: Doesn't this regex check for positive or negative floating point numbers?  Don't you just care about positive integers?

File: src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+
+    public String getSymbol() {
+        return this.symbol;
+    }
+
+    public String takeTurn(List<String> boardElements) {
+
+        String computerSymbol = getSymbol();
+        IndexScore minMaxValue = maximizer(boardElements, computerSymbol,
+                -1000000, 1000000, 1);
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java]: Put the initial values for AB pruning into constants

File: src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+
+        return minMaxValue.getIndex().toString();
+    }
+
+
+    public IndexScore maximizer(List<String> boardElements, String playerSymbol,
+                                int alpha, int beta, int depth) {
+
+        List<Integer> availableSpots = boardTools.getAvailableSpots(boardElements);
+        IndexScore maxScore = new IndexScore(1000, -1000);
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java]: Where do 1000 and -1000 come into play? I though we were using 1000000 and -1000000 above?

File: src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+        for (Integer move : availableSpots) {
+            boardElements.set(move, playerSymbol);
+
+            if (isTerminalState(boardElements)) {
+                maxScore = getTerminalScore(boardElements, move, depth);
+
+            } else {
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java]: You should definitely try to break this apart into smaller methods.

File: src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+        return maxScore;
+    }
+
+    public IndexScore minimizer(List<String> boardElements, String playerSymbol,
+                                int alpha, int beta, int depth) {
+
+        List<Integer> availableSpots = boardTools.getAvailableSpots(boardElements);
+        IndexScore minScore = new IndexScore(1000, 1000);
+
+        for (Integer move : availableSpots) {
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java]: Same thing as above - this method is gigantic. Try to break it up into smaller functions.

File: src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+    private boolean isTerminalState(List<String> boardElements) {
+        return boardTools.gameHasWinner(boardElements) || boardTools.gameIsTie(boardElements);
+    }
+
+    private IndexScore getTerminalScore(List<String> boardElements,
+                                        Integer boardIndex, int depth) {
+
+        if (boardTools.gameHasWinner(boardElements)) {
+            if (getSymbol().equals(boardTools.getWinner(boardElements))) {
+                return new IndexScore(boardIndex, 10 / depth);
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/ComputerPlayer.java]: Why are we using 10 and -10 here? You seem to have three different max/min values?

File: src/main/java/com/nexttran/tictactoecore/console/IOGameConsole.java [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        printMessage(boardRow + "\n");
+    }
+
+    private String makeBoardRow(String existingPart, String boardElement, int boardSize) {
+        if (boardSize == 3) {
+            return existingPart + boardElement + "   ";
+        }
+
+        if (boardSize > 3) {
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/console/IOGameConsole.java]: Didn't you hard-code the state of your element to have nine spaces (i.e. a board size of 3)? Do you need any of this code? Maybe I don't understand React as well as I thought?

File: src/main/java/com/nexttran/tictactoecore/console/IOPlayerConsole.java [hunk 25]
```diff
@@ -0,0 +1,35 @@
+package com.nexttran.tictactoecore.console;
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/console/IOPlayerConsole.java]: Have you tried IntelliJ. It will format your code quite nicely for you.

File: src/main/java/com/nexttran/tictactoecore/console/PlayTicTacToe.java [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+        TwoPlayerGameLoop ticTacToe;
+
+        if (startingPlayer.equals("1")) {
+            ticTacToe = new TwoPlayerGameLoop(humanPlayer, computerPlayer, gamePlay, iioGame);
+        }
+
+        else {
+            ticTacToe = new TwoPlayerGameLoop(computerPlayer, humanPlayer, gamePlay, iioGame);
+        }
+        ticTacToe.play(3);
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/console/PlayTicTacToe.java]: You `main` function is way too long - you want this business logic to be in classes that you define and put under test.

File: src/main/java/com/nexttran/tictactoecore/console/PlayTicTacToe.java [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+        System.out.println(promptMessage);
+
+        String playerInput = null;
+        try {
+            playerInput = br.readLine();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        return playerInput;
```
Comment (cmvandrevala) [inline src/main/java/com/nexttran/tictactoecore/console/PlayTicTacToe.java]: This would be where I expect the user to be able to select the board size for a console game.

File: src/test/java/com/nexttran/tictactoecore/web/springwithreact/BoardEntityControllersTests.java [hunk 28]
```diff
@@ -0,0 +1,176 @@
...
+
+                .andExpect(status().isOk())
+                .andExpect(content().json(stringBoard));
+    }
+
+    @Test
+    public void humanControllerShouldNOTmakeMoveAtSpotZeroAsIsAlreadyTaken() throws Exception {
+
+        board.setStatusCode(UNAVAILABLE_SPOT);
+        board.setZero("H");  // Computer move is C... So, H should not be replaced unless computer plays
```
Comment (cmvandrevala) [inline src/test/java/com/nexttran/tictactoecore/web/springwithreact/BoardEntityControllersTests.java]: Try to express these ideas without comments

**Comment by cmvandrevala** [comment]
This was a pretty massive PR. I assume that you were pulling a pre-existing project into a new Git Repo? I covered as much as I could in an hour, but I am happy to chat in person if you would like

**Comment by cmvandrevala** [comment]
I think I can distill my comments down to three bullet points:

1. YAGNI - You Ain't Gonna Need It, Don't write code or features that you don't directly need (e.g., the work you put into board size)
2. Code Hygiene - Use a linter and/or formatter to enforce a consistent style throughout your codebase. Consistency makes your code easier to follow.
3. I notice that you have a distinct, procedural style of programming. This is not a good or bad thing - some problems lend themselves to this style, others do not. You might want to look into OO and functional programming design (which might make writing minimax a bit easier too).

File: src/main/java/com/nexttran/tictactoecore/BoardTools.java [hunk 29]
```diff
@@ -0,0 +1,193 @@
...
+    public String getElement(List<String> boardElements, int index) {
+        return boardElements.get(index);
+    }
+
+    public int getBoardSize(List<String> boardElements) {
+        return (int) Math.sqrt((double) boardElements.size());
+    }
+
+    public boolean isSpotAvailable(List<String> boardElements, int index) {
+        int maxBoardIndex = (getBoardSize(boardElements)
```
Comment (rrugamba) [inline src/main/java/com/nexttran/tictactoecore/BoardTools.java]: - this function could be refactored to have 3 lines of code.
- there's an cleaner way to return actual boolean statements than using `return true` or `return false`.
    forexample line 24: 
     you could use `return isNumeric(boardElement)`
     instead of using `if (blah blah) return false` etc

File: src/main/java/com/nexttran/tictactoecore/BoardTools.java [hunk 30]
```diff
@@ -0,0 +1,193 @@
...
+        String boardElement = boardElements.get(index);
+        if (!(isNumeric(boardElement))) return false;
+        return true;
+    }
+
+    public List<Integer> getAvailableSpots(List<String> boardElements) {
+        List<Integer> availableSpots = new ArrayList<>();
+
+        for (int i = 0; i < boardElements.size(); i++) {
+            if (isSpotAvailable(boardElements, i)) availableSpots.add(i);
```
Comment (rrugamba) [inline src/main/java/com/nexttran/tictactoecore/BoardTools.java]: using a linter would help format this code a little better
something like this
`if(isSpotAvailable(boardElements, i))`
    `availableSpots.add(i)`
looks bit cleaner and easier to read

File: src/main/java/com/nexttran/tictactoecore/BoardTools.java [hunk 31]
```diff
@@ -0,0 +1,193 @@
...
+        if (gameHasWinner(boardElements)
+                || boardHasNonPlayedSpot(boardElements)) return false;
+        return true;
+    }
+
+    @Override
+    public boolean gameHasWinner(List<String> boardElements) {
+        int boardSize = getBoardSize(boardElements);
+
+        if (isHorizontalWin(boardElements, boardSize)) return true;
```
Comment (rrugamba) [inline src/main/java/com/nexttran/tictactoecore/BoardTools.java]: - again here
 you could just return a single statement using `||` and avoid using lots of `if`s and `return true` statements

- something like this:
`return (isHorizontalWin() || isVerticalWin() || isTopLeftDiagonalWin())` etc

File: src/main/java/com/nexttran/tictactoecore/console/TwoPlayerGameLoop.java [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+            if (!(boardElements.get(i).equals(newElements.get(i)))) {
+                play = boardElements.get(i);
+                break;
+            }
+        }
+        return play;
+    }
+
+    public String getErrorMessage(int statusCode) {
+        if (statusCode == 5) return "The game is over!";
```
Comment (rrugamba) [inline src/main/java/com/nexttran/tictactoecore/console/TwoPlayerGameLoop.java]: you could make good use of enums here

**Comment by kkrull** [comment]
I noticed in `pom.xml` that you're using the `maven-shade-plugin` in some cases.  I know it's super popular, and there are numerous sources that show it as a way to create a jar file with everything in it that is needed to run.

There is however a dark side, to using the Shade plugin.

Pay close attention to the warnings Maven is printing to the console when running this: there are multiple versions of the same class (who wins?), and the shaded, uber jar is overwriting the jar with just your code in it (causing issues for any future artifacts that depend upon this code).  There are [simpler alternatives](https://medium.com/@randilfernando/when-to-use-maven-jar-maven-assembly-or-maven-shade-ffc3f76ba7a6) that may be more appropriate, to this circumstance.  This is another case where everything seems to be working, until you one day when you [enter a world of pain](https://softwareengineering.stackexchange.com/a/351091/19043).

File: src/main/java/com/nexttran/tictactoecore/console/TwoPlayerGameLoop.java (hunk 32 above)
Comment (kkrull) [reply src/main/java/com/nexttran/tictactoecore/console/TwoPlayerGameLoop.java]: Expanding on Rugamba's comment - when you see if-else (or switch) blocks like this, it's often a sign that there's an [opportunity for greater polymorphism](https://sourcemaking.com/refactoring/smells/switch-statements).

Notice how there's a string associated with each numerical value?  That suggests that there may be a missing Status class, that keeps the code and its description together in one place.

File: src/main/java/com/nexttran/tictactoecore/web/springmvc/MvcConfiguration.java [hunk 33]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.web.servlet.config.annotation.EnableWebMvc;
+import org.springframework.web.servlet.config.annotation.ViewResolverRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+import org.springframework.web.servlet.view.InternalResourceViewResolver;
+import org.springframework.web.servlet.view.JstlView;
+
+@Configuration
+@EnableWebMvc
+@ComponentScan
+public class MvcConfiguration implements WebMvcConfigurer {
```
Comment (kkrull) [inline src/main/java/com/nexttran/tictactoecore/web/springmvc/MvcConfiguration.java]: I would recommend - sometime after this pull request - moving the web-related code to its own module, so that console players aren't forced to load a lot of Spring dependencies they don't need.  That said, I often see people add unnecessary complexity in the way that they break it up into multiple modules.  Great care must be taken, so I often recommend - as you have done here - to start with most of the code in a single module, at least until the core code stabilizes.

In other words, this is food for future thought; not an addressable finding for now.

File: src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebApplication.java [hunk 34]
```diff
@@ -0,0 +1,19 @@
+package com.nexttran.tictactoecore.web.springmvc;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.builder.SpringApplicationBuilder;
+import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
+
+@SpringBootApplication
```
Comment (kkrull) [inline src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebApplication.java]: I understand Spring documentation says to use Spring boot everywhere, because - when it works - it automates a lot of Spring-related work for you that is common to many projects.  However, I personally find it to be more complex than I am comfortable with.  I would consider questions such as: 1) how thoroughly I understand which version of which dependencies are being loaded, 2) where do I look for details on which classes are loaded from which jars, 3) where do I look for details on which classes are being instantiated, and injected as collaborators, and 4) do I feel comfortable debugging future class-loading issues?

Again this is one of those food for thought comments.  My only recommendation for the short term would be to spend some time reading Spring Boot's documentation and console output, so you have a better understanding of what it's doing before things go wrong.

File: src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebController.java [hunk 35]
```diff
@@ -0,0 +1,150 @@
...
+
+    private WebGame webGame =  new WebGame();
+    private GamePlay gamePlay = webGame.getGamePlay();
+    private BoardHelpers boardHelpers = new BoardHelpers();
+    private IPlayer webHumanPlayer;
+    private IPlayer computerPlayer = new ComputerPlayer("C", new BoardTools());
+    private final List<String> initialBoard = webGame.getInitialBoard(3);
+    private List<String> boardElements;
+    private int statusCode;
+    private MoveResult moveResult;
```
Comment (kkrull) [inline src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebController.java]: Although it may not be causing an issue right now, you can further separate concerns and avoid [potential concurrency issues with multiple instances of this controller](https://stackoverflow.com/a/11508436/112682) by moving any state-related code out to another class.  That said, the way you are using cookies is an idiomatic way to track state using a stateless HTTP protocol, as long as the state remains small enough to transfer on each request and response.

File: src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebController.java [hunk 36]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    @RequestMapping(value = "/computer/start", method = RequestMethod.GET)
+    public ModelAndView computerStart(HttpServletResponse response, ModelMap model) {
+
+        moveResult = gamePlay.playOnce(initialBoard, computerPlayer);
+        statusCode = moveResult.getStatusCode();
+        boardElements = moveResult.getBoardElements();
+
+        model = boardHelpers.addElementsOnModel(model, boardElements);
```
Comment (kkrull) [inline src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebController.java]: The re-assignment here doesn't appear to be necessary, and [won't be visible outside of this method](https://stackoverflow.com/a/40523/112682) anyway.

File: src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebController.java [hunk 37]
```diff
@@ -0,0 +1,150 @@
...
+        boardHelpers.createCookieAndAddToResponse(boardElements, response);
+
+        if (!(gamePlay.gameIsOngoing(boardElements))) {
+            return announceGameEnding(boardElements, model);
+
+        } else if (statusCode == AVAILABLE_SPOT) {
```
Comment (kkrull) [inline src/main/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebController.java]: Moving this check of the status code to a new `MoveResult#isAvailableSpot` method would better separate the meaning of the status (a space is available) from its _representation_ as data (the constant -1).  If MoveResult#getStatusCode chose one day to return some other value to indicate an available spot, this code would need to be updated (and it would be easy to not notice, unless a test failed).

File: src/main/java/com/nexttran/tictactoecore/web/springmvc/WebBoard.java [hunk 38]
```diff
@@ -0,0 +1,95 @@
+package com.nexttran.tictactoecore.web.springmvc;
+
+
+public class WebBoard {
```
Comment (kkrull) [inline src/main/java/com/nexttran/tictactoecore/web/springmvc/WebBoard.java]: It's good that you're separating the web-based code from the core Tic Tac Toe logic and from the console-based code.  As you move further down this path of having separate modules for separate ways of running the game, you may notice one of these potential changes: 1) this class might be deleted entirely (there's nothing strictly web-related in it, right now), 2) this class may extend the base TicTacToeBoard class in a way that is needed to work with Spring MVC (adding the getters and setters, for example, but retaining the core logic already in the core Board class)

File: src/main/java/com/nexttran/tictactoecore/web/springmvc/WebBoard.java [hunk 39]
```diff
@@ -0,0 +1,95 @@
...
+    private String five;
+    private String six;
+    private String seven;
+    private String eight;
+
+    public WebBoard() {
+        super();
+    }
+
+    public String getZero() {
```
Comment (kkrull) [inline src/main/java/com/nexttran/tictactoecore/web/springmvc/WebBoard.java]: I imagine Spring is probably using these getters and setters behind the scenes to set the value of each space, when players take turns.  Often times, a class with only getters and setters is a sign of [feature envy](https://sourcemaking.com/refactoring/smells/feature-envy) and an opportunity to add methods to this class, that operate upon the data and produce a result (or side-effect).  For example, what if you added WebBoard#createCookie to do what `BoardHelpers#createCookies` is doing?  You may not need so many getters and setters, then (although Spring may force you to retain them, but I see that as a separate issue from general encapsulation).

File: src/main/webapp/WEB-INF/view/showBoard.jsp [hunk 40]
```diff
@@ -0,0 +1,68 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
```
Comment (kkrull) [inline src/main/webapp/WEB-INF/view/showBoard.jsp]: This taglib doesn't appear to be in use right now.  I'd suggest deleting the reference until it's needed.

File: src/test/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebControllersTest.java [hunk 41]
```diff
@@ -0,0 +1,150 @@
...
+
+        Cookie cookie = new Cookie("board", "0_1_2_C_C_C_6_7_8");
+        mockMvc.perform(post("/human/play/").param("zero", "0").cookie(cookie))
+                .andExpect(status().isOk())
+                .andExpect(cookie().value("board", "0_1_2_C_C_C_6_7_8"));
+
+    }
+
+    @Test
+    public void ifGameIsOverComputerShouldNotBeAllowedToPlay() throws Exception {
```
Comment (kkrull) [inline src/test/java/com/nexttran/tictactoecore/web/springmvc/TicTacToeWebControllersTest.java]: Overall these tests seem to cover the controller fairly well, and it's particularly helpful that you named the test methods in a way that describes the intended behavior (despite that being hard to do in JUnit 4).  Take a closer look at how the name of the test - "if the game is over" - matches to the test data.  You create a game board that _happens_ to represent a game over state, but what if you made an abstraction to simply say whether the game is over or not?  That would simplify the logic in your controller and make it easier to test game state transitions and web controller methods separately.

**Comment by kkrull** [comment]
I can see that you've been working hard on this, and there's a lot of feedback in this PR.  I hope that you will find it useful, but I understand that it can also be overwhelming and cause a feature to be blocked at times.

With that in mind, I wouldn't delay merging this just to satisfy all the comments I've made.  Instead, I'd suggest just to do a bit of research (to understand what is being suggested) for now, and keeping these comments in mind as you do future work.  It may make sense to try some of the suggestions in a future PR as a separate effort, or you might be working in those files again someday and make some changes while you're doing that work.

File: reactapp/src/BoardComponent.js (hunk 12 above)
Comment (ssunday) [reply reactapp/src/BoardComponent.js]: ^ I was going to jump on that and say that if you're not using enums, maybe make some extracted logic that handles these codes and checks what state they are? So it's clearer?
